src/cargo/ops/cargo_update.rs

```diff
@@ -165,7 +182,13 @@ pub fn update_lockfile(ws: &Workspace<'_>, opts: &UpdateOptions<'_>) -> CargoRes
         .filter(|s| !s.is_registry())
         .collect();
 
-    let keep = |p: &PackageId| !to_avoid_sources.contains(&p.source_id()) && !to_avoid.contains(p);
+    let keep = |p: &PackageId| {
+        (!to_avoid_sources.contains(&p.source_id()) && !to_avoid.contains(p))
+                // In case of `--breaking` (or precise upgrades, coming in
+                // #14140), we want to keep all packages unchanged that didn't
+                // get upgraded.
+                    || (opts.breaking && !upgrades.contains_key(&(p.name(), p.source_id())))
```

Comment:
This is an important change that has been discussed here: https://github.com/rust-lang/cargo/pull/14140#discussion_r1676447754

Reply:
To put another way, there is concern with using `keep` like this and resolving this is a blocker for moving forward.

Reply:
I am curious what the actual concern is, though. We are only modifying `keep` in the case of breaking or precise updates, and we have quite extensive test coverage of those.

Reply:
> The reason this is not put in to_avoid is that those require PackageIds, i.e. they require a version. When doing upgrades, we don't have any versions, only version requirements.

In general converting from a requirement to a version is not possible and converting to a set of versions requires scanning  all the versions on the index. But in this case `keep`  is only called on versions that are in the lock file. I think we already scan all the versions in the lock file to construct `to_avoid`, so I'd rather see this check  done there.

Reply:
I have changed this PR so it no longer changes `keep`, but instead builds up what it needs in `to_avoid`. Please review!